distributed/comm/ucx.py

```diff
@@ -0,0 +1,336 @@
+
+
+def _unparse_host_port(host, port=None):
+    return unparse_host_port(host, port)
+
+
+def get_endpoint_address(endpoint):
+    # TODO: ucx-py: 18
+    pass
+
+
+# ----------------------------------------------------------------------------
+# Comm Interface
+# ----------------------------------------------------------------------------
+
+
+class UCX(Comm):
+    """Comm object using UCP.
+
+    Parameters
+    ----------
+    ep : ucp.ucp_py_ep
+        The UCP endpoint.
+    address : str
+        The address, prefixed with `ucx://` to use.
+    deserialize : bool, default True
+        Whether to deserialize data in :meth:`distributed.protocol.loads`
+
+    Notes
+    -----
+    The read-write cycle uses the following pattern:
+
+    Each msg is serialized into a number of "data" frames. We prepend these
+    real frames with two additional frames
+
+        1. is_gpu: Boolean indicator for whether the frame should be
+           received into GPU memory. Packed in '?' format. Unpack with
+           ``<n_frames>?`` format.
+        2. frame_size : Unisigned int describing the size of frame (in bytes)
+           to receive. Packed in 'Q' format, so a length-0 frame is equivalent
+           to an unsized frame. Unpacked with ``<n_frames>Q``.
+
+    The expected read cycle is
+
+    1. Read the frame describing number of frames
+    2. Read the frame describing whether each data frame is gpu-bound
+    3. Read the frame describing whether each data frame is sized
+    4. Read all the data frames.
+    """
+
+    def __init__(self, ep: ucp.ucp_py_ep,
+                 address: str,
+                 listener_instance,
+                 deserialize=True):
+        logger.info("UCX.__init__ %s %s", address, listener_instance)
+        self.ep = ep
+        assert address.startswith("ucx")
+        self.address = address
+        self.listener_instance = listener_instance
+        default_port = next(_PORT_COUNTER)
+        self._host, self._port = _parse_host_port(address, default_port)
+        self._local_addr = None
+        self.deserialize = deserialize
+
+        # finalizer?
+
+    @property
+    def local_address(self) -> str:
+        return self._local_addr
+
+    @property
+    def peer_address(self) -> str:
+        # XXX: This isn't quite for the server (from UCXListener).
+        # We need the port? Or the tag?
+        return self.address
+
+    async def write(self, msg: dict, serializers=None, on_error: str = "message"):
+        frames = await to_frames(
+            msg, serializers=serializers, on_error=on_error
+        )  # TODO: context=
+        gpu_frames = b''.join([struct.pack("?", hasattr(frame, '__cuda_array_interface__'))
+                               for frame in frames])
+        size_frames = b''.join([struct.pack("Q", nbytes(frame)) for frame in frames])
+
+        frames = [gpu_frames] + [size_frames] + frames
+        nframes = struct.pack("Q", len(frames))
```

Comment:
I wouldn't mind avoiding the suffix `_frames` above for `gpu_frames` and `size_frames`.  I typically think of a frame as a large buffer that we're sending.  These appear to be more metadata.  Perhaps `is_gpus` and sizes` or something similar?